[AMDGPU] Src1 of VOP3 DPP instructions can be SGPR on supported subtargets

mbrkusanin opened the pull request:
In order to avoid duplicating every dpp pseudo opcode that has src1, we allow it for all opcodes and add manual checks on subtargets that do not support it.

rampitec commented on llvm/test/CodeGen/AMDGPU/dpp_combine_gfx11.mir:
```diff
@@ -28,11 +30,53 @@ body:             |
     %11:vgpr_32 = V_MOV_B32_dpp %3, %1, 1, 15, 15, 1, implicit $exec
     %12:vgpr_32 = V_MED3_F32_e64 0, %11, 0, 2, 0, %7, 0, 0, implicit $mode, implicit $exec
 ...
 
```
Missing '---' between tests. Here and after the function.

rampitec commented on llvm/test/MC/AMDGPU/gfx1150_asm_features.s:
```diff
@@ -0,0 +1,20 @@
+
```
Remove empty line.

Sisyph reviewed (commented):
Please add a disassembler test. It might be permissible to accept sgpr as src1 on gfx11 as well as gfx1150 (i.e. disassembler is not strict), but we should at least ensure it is acceptable on gfx1150.

rampitec reviewed (commented):
LGTM with couple nits.

mbrkusanin committed:
[AMDGPU] Src1 of VOP3 DPP instructions can be SGPR on supported subtargets

In order to avoid duplicating every dpp pseudo opcode that has src1, we
allow it for all opcodes and add manual checks on subtargets that do not
support it.

mbrkusanin commented on llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp:
```diff
@@ -301,7 +302,14 @@ MachineInstr *GCNDPPCombine::createDPPInst(MachineInstr &OrigMI,
+      int OpNum = NumOperands;
+      // If subtarget does not support SGPRs for src1 operand then the
+      // requirements are the same as for src0. We check src0 instead because
+      // pseudos are shared between subtargets and allow SGPR for src1 on all.
```
I've added the assert for size but I'm not sure what float/not float match would be.

Sisyph replied:
Something like assert(src0->getType()->isFloatTy() == src1)->getType()->isFloatTy()) ? Do you think it's useful?

mbrkusanin replied:
But we do not have that info this late. Only reg class and size. We would have to look at each opcode separately. Besides, isOperandLegal (which is what we are lying to here) does not seem to care about int/float.

Sisyph reviewed (approved):
LGTM

jayfoad commented:
@mbrkusanin @Sisyph it seems like this works for integer instructions:
```
$ llvm-mc -arch=amdgcn -mcpu=gfx1150 -show-encoding <<< "v_add_nc_u32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]"
	.text
	v_add_nc_u32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0] row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x00,0x25,0xd5,0xfa,0x06,0x00,0x00,0x01,0x1b,0x00,0xff]
```
but not for floating point:
```
$ llvm-mc -arch=amdgcn -mcpu=gfx1150 -show-encoding <<< "v_add_f32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]"
	.text
<stdin>:1:27: error: invalid operand for instruction
v_add_f32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]
                          ^
```
Is that a bug or is there some reason for it?

mbrkusanin commented:
> @mbrkusanin @Sisyph it seems like this works for integer instructions:
> 
> ```
> $ llvm-mc -arch=amdgcn -mcpu=gfx1150 -show-encoding <<< "v_add_nc_u32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]"
> 	.text
> 	v_add_nc_u32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0] row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x00,0x25,0xd5,0xfa,0x06,0x00,0x00,0x01,0x1b,0x00,0xff]
> ```
> 
> but not for floating point:
> 
> ```
> $ llvm-mc -arch=amdgcn -mcpu=gfx1150 -show-encoding <<< "v_add_f32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]"
> 	.text
> <stdin>:1:27: error: invalid operand for instruction
> v_add_f32_e64_dpp v5, v1, s3 quad_perm:[3,2,1,0]
>                           ^
> ```
> 
> Is that a bug or is there some reason for it?

A bug. Looks like class for src1_modifiers needs to be updated in tablegen, it differs for ints and floats. I'm working on a patch.
